File: c_implementation/reaction_wheel_actuator.c
#include "reaction_wheel_actuator.h"

static double clamp_value(
    double value,
    double minimum,
    double maximum
)
{
    if (value > maximum)
    {
        return maximum;
    }

    if (value < minimum)
    {
        return minimum;
    }

    return value;
}

void compute_reaction_wheel_actuation(
    const double requested_spacecraft_torque[3],
    const double wheel_speed[3],
    const double wheel_inertia[3],
    const double wheel_torque_limit[3],
    const double wheel_speed_limit[3],
    double applied_spacecraft_torque[3],
    double wheel_acceleration[3],
    double wheel_motor_torque[3],
    int speed_limited[3]
)
{
    for (int axis = 0; axis < 3; axis++)
    {
        /*
         * First apply the wheel motor-torque limit to the requested
         * spacecraft torque.
         */
        applied_spacecraft_torque[axis] =
            clamp_value(
                requested_spacecraft_torque[axis],
                -wheel_torque_limit[axis],
                wheel_torque_limit[axis]
            );

        /*
         * The wheel motor torque is equal and opposite to the
         * spacecraft torque.
         */
        wheel_motor_torque[axis] =
            -applied_spacecraft_torque[axis];

        speed_limited[axis] = 0;

        /*
         * Prevent a wheel at its positive speed limit from receiving
         * additional positive wheel torque.
         */
        if (
            wheel_speed[axis] >= wheel_speed_limit[axis]
            && wheel_motor_torque[axis] > 0.0
        )
        {
            wheel_motor_torque[axis] = 0.0;
            speed_limited[axis] = 1;
        }

        /*
         * Prevent a wheel at its negative speed limit from receiving
         * additional negative wheel torque.
         */
        if (
            wheel_speed[axis] <= -wheel_speed_limit[axis]
            && wheel_motor_torque[axis] < 0.0
        )
        {
            wheel_motor_torque[axis] = 0.0;
            speed_limited[axis] = 1;
        }

        applied_spacecraft_torque[axis] =
            -wheel_motor_torque[axis];

        if (wheel_inertia[axis] > 0.0)
        {
            wheel_acceleration[axis] =
                wheel_motor_torque[axis]
                / wheel_inertia[axis];
        }
        else
        {
            wheel_acceleration[axis] = 0.0;
        }
    }
}

File: attitude_control/reaction_wheel_actuator.m
function [spacecraft_torque, wheel_acceleration, ...
    wheel_motor_torque, speed_limited] = ...
    reaction_wheel_actuator( ...
        requested_spacecraft_torque, ...
        wheel_speed, ...
        wheel_inertia, ...
        wheel_torque_max, ...
        wheel_speed_max)
%REACTION_WHEEL_ACTUATOR Three orthogonal reaction-wheel model.
%
% Inputs:
%   requested_spacecraft_torque - controller request in N m
%   wheel_speed                 - wheel speeds in rad/s
%   wheel_inertia               - rotor inertias in kg m^2
%   wheel_torque_max            - motor torque limits in N m
%   wheel_speed_max             - wheel speed limits in rad/s
%
% Outputs:
%   spacecraft_torque - torque applied to spacecraft in N m
%   wheel_acceleration - wheel acceleration in rad/s^2
%   wheel_motor_torque - torque applied to each wheel in N m
%   speed_limited      - logical flag for each wheel

%% Equal-and-opposite wheel torque request
requested_wheel_torque = -requested_spacecraft_torque;

%% Apply motor torque limits
wheel_motor_torque = max( ...
    min(requested_wheel_torque, wheel_torque_max), ...
    -wheel_torque_max);

%% Prevent acceleration farther beyond the speed limit
speed_limited = false(3,1);

for axis = 1:3
    at_positive_limit = ...
        wheel_speed(axis) >= wheel_speed_max(axis);

    at_negative_limit = ...
        wheel_speed(axis) <= -wheel_speed_max(axis);

    accelerating_positive = wheel_motor_torque(axis) > 0;
    accelerating_negative = wheel_motor_torque(axis) < 0;

    if (at_positive_limit && accelerating_positive) || ...
       (at_negative_limit && accelerating_negative)

        wheel_motor_torque(axis) = 0;
        speed_limited(axis) = true;
    end
end

%% Wheel acceleration
wheel_acceleration = ...
    wheel_motor_torque ./ wheel_inertia;

%% Equal-and-opposite spacecraft torque
spacecraft_torque = -wheel_motor_torque;

end